crates/swc_ecma_transforms_react/src/refresh/hook.rs

```diff
@@ -0,0 +1,523 @@
+                self.gen_hook_register_stmt(f.ident.clone(), sig);
+            }
+        }
+    }
+}
+
+fn collect_hooks(stmts: &mut Vec<Stmt>, cm: &SourceMap) -> Option<HookSig> {
+    let mut hook = HookCollector {
+        state: Vec::new(),
+        cm,
+    };
+
+    stmts.visit_with(&mut hook);
+
+    if hook.state.len() > 0 {
+        let sig = HookSig::new(hook.state);
+        stmts.insert(0, make_call_stmt(sig.handle.clone()));
+
+        Some(sig)
+    } else {
+        None
+    }
+}
+
+fn collect_hooks_arrow(body: &mut BlockStmtOrExpr, cm: &SourceMap) -> Option<HookSig> {
+    match body {
+        BlockStmtOrExpr::BlockStmt(block) => collect_hooks(&mut block.stmts, cm),
+        BlockStmtOrExpr::Expr(expr) => {
+            let mut hook = HookCollector {
+                state: Vec::new(),
+                cm,
+            };
+
+            expr.visit_with(&mut hook);
+
+            if hook.state.len() > 0 {
+                let sig = HookSig::new(hook.state);
+                *body = BlockStmtOrExpr::BlockStmt(BlockStmt {
+                    span: expr.span(),
+                    stmts: vec![
+                        make_call_stmt(sig.handle.clone()),
+                        Stmt::Return(ReturnStmt {
+                            span: expr.span(),
+                            arg: Some(Box::new(expr.as_mut().take())),
+                        }),
+                    ],
+                });
+                Some(sig)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+struct HookCollector<'a> {
+    state: Vec<Hook>,
+    cm: &'a SourceMap,
+}
+
+static IS_HOOK_LIKE: Lazy<Regex> = Lazy::new(|| Regex::new("^use[A-Z]").unwrap());
+impl<'a> HookCollector<'a> {
+    fn get_hook_from_call_expr(&self, expr: &CallExpr, lhs: Option<&Pat>) -> Option<Hook> {
+        let callee = if let ExprOrSuper::Expr(callee) = &expr.callee {
+            Some(callee.as_ref())
+        } else {
+            None
+        }?;
+        let mut hook_call = None;
+        let ident = match callee {
+            Expr::Ident(ident) => {
+                hook_call = Some(HookCall::Ident(ident.clone()));
+                Some(ident)
+            }
+            Expr::Member(MemberExpr {
+                obj: ExprOrSuper::Expr(obj),
+                prop,
+                ..
+            }) => {
+                if let Expr::Ident(ident) = prop.as_ref() {
+                    hook_call = Some(HookCall::Member(*obj.clone(), ident.clone()));
+                    Some(ident)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }?;
+        let name = if IS_HOOK_LIKE.is_match(&ident.sym) {
+            Some(ident)
+        } else {
+            None
+        }?;
+        let mut key = if let Some(name) = lhs {
+            self.cm
+                .span_to_snippet(name.span())
+                .unwrap_or("".to_string())
```

Comment:
I think we should use `unwrap_or_else` or `.unwrap_or(String::new())`.

Maybe specialization used by std may optimize this, but just to be safe.